### DISCUSSION TITLE
Release cycle?!?

### DISCUSSION THREAD

**Discussion by jedie** [beartype/beartype#433] [Q&A] [answered]
There are many, many commits since last release: https://github.com/beartype/beartype/compare/v0.18.5...main

What's the release cycle? Seems it's not:

> Release early, release often

isn't it?

**Comment by JWCS** [comment, marked as answer]
In general, it's a release when there's enough stable new stuff. There were two pre-release candidates; `pip install --upgrade --pre beartype`, over the last 3-4 months, and v0.19 has been almost coming out... until some ~pesky~ insightful user finds a bug, or a new feature needs more polishing before being released to the wild. 
In general, the project is 1 guy in volunteer time, chipping away at issues; so if you want something sooner than --pre, it has to be `pip install https://github.com/beartype/beartype/archive/main.zip`, for various fluctuating definitions of stable. 
PRs welcome. Oh, and as for the many commits... it was a big new feature; just one, but very big.

**Comment by leycec** [comment]
**Big bear hugs to @JWCS.** As always, @JWCS has the right of it. @beartype 0.19.0 is currently feature-complete and *almost* – but not quite – bug-free. My Five Month Plan™ was to push out @beartype 0.19.0 last Friday. I swear! Would @leycec lie? :face_with_open_eyes_and_hand_over_mouth:

Tragically, @minmax discovered [last-minute critical issue #433](https://github.com/beartype/beartype/discussions/433) at the intersection of `**kwargs` and `typing.TypeDict` subclasses. Specifically, PEP 692-compliant type hints resembling `**kwargs: typing.Unpack[typing.TypedDict]` currently cause @beartype to raise unreadable exceptions. Unacceptable, @beartype!

Last Saturday evening, <sup>*RIP @leycec's fun time*</sup> I *probably* resolved #433 with commit c6a3f1acb214e. But I haven't actually tested anything with unit tests yet. So, everything is almost certainly still broken. I hope to have everything fully tested and resolved by this Friday. Which then leads us to...

## A New @beartype `0.19.0rc2` Beta Release!

**Oh, yes.** That's right. You heard the horrifying news here first. I'll be releasing the last (and hopefully final) beta pre-release of this endless `0.19.0` dev cycle that everyone hates. Even our cats hate this dev cycle.

If my *wonderful* users give the "This doesn't suck as bad as 0.18.0!" thumbs-up, I'll then release a few days later...

## A Stable @beartype `0.19.0` Full Release by September 30th

**Oh, yes.** That's September 30th of *this* year, by the way. I swear! Which leads us to...

## Users Hate These 5 @beartype Release Tricks

I get it. @beartype is absolutely *not* adhering to this common-sense project management credo:

> Release early, release often

For ordinary packages, this is the best possible advice. For @beartype, this is the worst possible advice. Why? Because **@beartype is mission-critical.** When @beartype breaks, increasingly the entire Python ecosystem breaks. This includes PyTorch – which then transitively includes ChatGPT, OpenAI, Microsoft, and by extension the entirety of American late-stage capitalism. Do we grok the stakes here? The stakes somehow become a *whole* lot larger than "one bald autist has fun smashing code together in a remote Canadian cottage."

I should probably be paid to do this immense and insane thing we call @beartype. Imagine if all neurosurgeons were unpaid volunteers. This is hyperbole, but it's also not. @beartype is the neurosurgeon that fixes bugs during LLM training. Much like Soviets under the USSR, I just *pretend* that I'm being paid by behaving responsibly towards the rest of humanity. "Release early, release often" is what I *used* to believe. Then I broke PyTorch with the ill-fated @beartype 0.18.0 release. The heavens themselves unleashed Hellish Displeasure. Now, I choose wisely.

The new motto is:

> Release late. Release infrequently. Release safely.

On the bright side, "Release safely." is good! We can all agree. On the dark side, "Release late." and "Release infrequently." are both bad. We still agree. But one out of two ain't bad. Right?

![](https://media1.tenor.com/m/7AFIOKFvE3UAAAAd/i-might-not-have-all-the-things-i-love-in-my-life-but-two-out-of-three-ain%27t-bad.gif)
<sup>*that feeling when you realize it isn't even two out of three*</sup>
